tests/support/client.rs

```diff
@@ -249,36 +294,129 @@ impl hyper::client::connect::Connect for Conn {
+    },
+    Handshake {
+        future: tokio_rustls::Connect<TcpStream>,
+        running: Arc<Mutex<Option<oneshot::Sender<()>>>>,
+    },
 }
 
-impl io::Read for RunningIo {
+impl Future for ConnectorFuture {
+    type Item = RunningIo;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let take_running = move |r: &Mutex<Option<oneshot::Sender<()>>>| {
+            r.lock()
+                .expect("running lock")
+                .take()
+                .expect("support client cannot connect more than once")
+        };
+
+        loop {
+            *self = match self {
+                ConnectorFuture::Init {
+                    future,
+                    tls,
+                    running,
+                } => {
+                    let io = try_ready!(future.poll());
+
+                    match tls {
+                        None => {
+                            return Ok(Async::Ready(RunningIo::Http(io, take_running(running))));
+                        }
+
+                        Some(TlsConfig {
+                            client_config,
+                            name,
+                        }) => {
+                            let future = tokio_rustls::TlsConnector::from(client_config.clone())
+                                .connect(DNSName::as_ref(name), io);
+                            ConnectorFuture::Handshake {
+                                future,
+                                running: running.clone(),
+                            }
+                        }
+                    }
+                }
+
+                ConnectorFuture::Handshake { future, running } => {
+                    let io = try_ready!(future.poll());
+                    return Ok(Async::Ready(RunningIo::Https(io, take_running(running))));
+                }
+            }
+        }
+    }
+}
+
+enum RunningIo {
+    Http(TcpStream, oneshot::Sender<()>),
+    Https(TlsStream<TcpStream, ClientSession>, oneshot::Sender<()>),
+}
+
+impl Read for RunningIo {
+    #[inline]
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        self.inner.read(buf)
+        match *self {
+            RunningIo::Http(ref mut s, _) => s.read(buf),
+            RunningIo::Https(ref mut s, _) => s.read(buf),
+        }
     }
 }
 
-impl io::Write for RunningIo {
+impl Write for RunningIo {
+    #[inline]
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        self.inner.write(buf)
+        match *self {
+            RunningIo::Http(ref mut s, _) => s.write(buf),
+            RunningIo::Https(ref mut s, _) => s.write(buf),
+        }
     }
 
+    #[inline]
     fn flush(&mut self) -> io::Result<()> {
-        self.inner.flush()
+        match *self {
+            RunningIo::Http(ref mut s, _) => s.flush(),
+            RunningIo::Https(ref mut s, _) => s.flush(),
+        }
     }
 }
 
-impl AsyncRead for RunningIo {}
+impl AsyncRead for RunningIo {
```

Comment:
Unless you had a reason to add this, we should still be okay with `impl AsyncRead for Running IO {}`. Removing this I was still able to run all the tests, but let me know if you were thinking it will be helpful for certain future tests!